khal/controllers.py

```diff
@@ -528,27 +528,28 @@ def import_ics(collection, conf, ics, batch=False, random_uid=False, format=None
                   even when an event with the same uid already exists
     :type batch: bool
     """
-    cal = icalendar.Calendar.from_ical(ics)
-    events = [item for item in cal.walk() if item.name == 'VEVENT']
-    events_grouped = defaultdict(list)
-    for event in events:
-        events_grouped[event['UID']].append(event)
-
     if format is None:
         format = conf['view']['event_format']
 
-    vevents = list()
-    for uid in events_grouped:
-        vevents.append(sorted(events_grouped[uid], key=sort_key))
+    vevents = utils.split_ics(ics)
+
     for vevent in vevents:
-        import_event(vevent, collection, conf['locale'], batch, random_uid, format, env)
+        import_event(
+            vevent, collection, conf['locale'], batch, random_uid, format, env,
+        )
 
 
 def import_event(vevent, collection, locale, batch, random_uid, format=None, env=None):
-    """import one event into collection, let user choose the collection"""
+    """import one event into collection, let user choose the collection
 
+    :type vevent: list of vevents, which can be more than one VEVENT, i.e., the
+        same UID, i.e., one "master" event and (optionally) 1+ RECURRENCE-ID events
+    :type vevent: list(str)
+    """
+    # TODO re-enable random_uid
     # print all sub-events
-    for sub_event in vevent:
+    cal = icalendar.Calendar.from_ical(vevent)
+    for sub_event in [item for item in cal.walk() if item.name == 'VEVENT']:
```

Comment:
Using `()` instead of `[]` would at least make it a generator comprehension instead of listcomp, and have the same effect of skipping creation of an intermediate list.